Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs

```diff
@@ -0,0 +1,327 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.NodejsTools.Npm;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.NodejsTools.NpmUI
+{
+    internal sealed class NpmWorker : IDisposable
+    {
+        private static readonly Uri defaultRegistryUri = new Uri("https://registry.npmjs.org/");
+
+        private readonly INpmController npmController;
+        private readonly Queue<QueuedNpmCommandInfo> commandQueue = new Queue<QueuedNpmCommandInfo>();
+        private readonly object queuelock = new object();
+
+        private bool isDisposed;
+        private bool isExecutingCommand;
+        private readonly Thread worker;
+        private QueuedNpmCommandInfo currentCommand;
+        private INpmCommander commander;
+
+        public NpmWorker(INpmController controller)
+        {
+            this.npmController = controller;
+
+            this.worker = new Thread(this.Run)
+            {
+                Name = "npm worker Execution",
+                IsBackground = true
+            };
+            this.worker.Start();
+        }
+
+        private void Pulse()
+        {
+            lock (this.queuelock)
+            {
+                Monitor.PulseAll(this.queuelock);
+            }
+        }
+
+        public bool IsExecutingCommand
+        {
+            get
+            {
+                lock (this.queuelock)
+                {
+                    return this.isExecutingCommand;
+                }
+            }
+            set
+            {
+                lock (this.queuelock)
+                {
+                    this.isExecutingCommand = value;
+                    Pulse();
+                }
+            }
+        }
+
+        private void QueueCommand(QueuedNpmCommandInfo info)
+        {
+            lock (this.queuelock)
+            {
+                if (this.commandQueue.Contains(info)
+                    || info.Equals(this.currentCommand))
+                {
+                    return;
+                }
+                this.commandQueue.Enqueue(info);
+                Monitor.PulseAll(this.queuelock);
+            }
+        }
+
+        public void QueueCommand(string arguments)
+        {
+            QueueCommand(new QueuedNpmCommandInfo(arguments));
+        }
+
+        private async void Execute(QueuedNpmCommandInfo info)
+        {
+            this.IsExecutingCommand = true;
+            INpmCommander cmdr = null;
+            try
+            {
+                lock (this.queuelock)
+                {
+                    cmdr = this.npmController.CreateNpmCommander();
+
+                    this.commander = cmdr;
+                }
+
+                await cmdr.ExecuteNpmCommandAsync(info.Arguments);
+            }
+            finally
```

Comment:
This function could interleave with itself and cause `commander` / `isExecutingCommand` to be out of sync with each other. Writes to those fields need to be in the same lock block.